ssl/record/methods/dtls_meth.c

```diff
@@ -461,25 +492,106 @@ int dtls_get_more_records(OSSL_RECORD_LAYER *rl)
 
         rl->rstate = SSL_ST_READ_BODY;
 
-        p = rl->packet;
+        if (!PACKET_buf_init(&dtlsrecord, rl->packet, rl->packet_length)
+            || !PACKET_get_1(&dtlsrecord, &record_type)) {
+            rl->packet_length = 0;
+            goto again;
+        }
 
         /* Pull apart the header into the DTLS1_RECORD */
-        rr->type = *(p++);
-        ssl_major = *(p++);
-        ssl_minor = *(p++);
-        rr->rec_version = (ssl_major << 8) | ssl_minor;
+        rr->type = (int)record_type;
+
+        /*-
+         * rfc9147:
+         * Implementations can demultiplex DTLS 1.3 records by examining the first
+         * byte as follows:
+         *   • If the first byte is alert(21), handshake(22), or ack(proposed, 26),
+         *     the record MUST be interpreted as a DTLSPlaintext record.
+         *   • If the first byte is any other value, then receivers MUST check to
+         *     see if the leading bits of the first byte are 001. If so, the implementation
+         *     MUST process the record as DTLSCiphertext; the true content type
+         *     will be inside the protected portion.
+         *   • Otherwise, the record MUST be rejected as if it had failed deprotection,
+         *     as described in Section 4.5.2.
+         */
+        if (rl->version == DTLS1_3_VERSION
+            && rr->type != SSL3_RT_ALERT
+            && rr->type != SSL3_RT_HANDSHAKE
+            /* TODO(DTLSv1.3): && rr->type != SSL3_RT_ACK depends on acknowledge implementation */
+            && !DTLS13_UNI_HDR_FIX_BITS_IS_SET(rr->type)) {
+            /* Silently discard*/
+            rr->length = 0;
+            rl->packet_length = 0;
+            goto again;
+        }
+
+        if (DTLS13_UNI_HDR_FIX_BITS_IS_SET(rr->type)) {
+            /*
+             * rfc9147:
+             * receivers MUST check to if the leading bits of the first byte are 001.
+             * If so, the implementation MUST process the record as DTLSCiphertext;
+             */
+            int cbitisset = DTLS13_UNI_HDR_CID_BIT_IS_SET(rr->type);
+            int sbitisset = DTLS13_UNI_HDR_SEQ_BIT_IS_SET(rr->type);
+            int lbitisset = DTLS13_UNI_HDR_LEN_BIT_IS_SET(rr->type);
+            uint16_t eebits = rr->type & DTLS13_UNI_HDR_EPOCH_BITS_MASK;
+
+            record_version = DTLS1_2_VERSION;
+            epoch = rl->epoch;
+            recseqnumlen = sbitisset ? 2 : 1;
+            recseqnumoffs = sizeof(recseqnum) - recseqnumlen;
+
+            if (/* OpenSSL does not support connection ID's and a connection ID
+                 * is only set if one has been negotiated: silently discard */
```

Comment:
Nit: no apostrophe in ID's and please reformat properly or just have `/*  OpenSSL does not support connection IDs: silently discard */`